### PR TITLE
Improve error message on trailing path slashes

### LINKED ISSUE #1138: "callPackage /dir/with-trailing-slash/ {}" throws misleading error
(Moved from https://github.com/NixOS/nixpkgs/issues/20621.)

## Issue description
```
$ nix-repl '<nixpkgs>'
nix-repl> pkgs.callPackage /tmp/mytest/ {}
error: value is a set while an integer was expected, at (string):1:1
```

By removing the trailing slash, everything is fine:

```
nix-repl> pkgs.callPackage /tmp/mytest {}
«derivation /nix/store/6wyvf5wwx2cxic6ydxgaky6mq22m2c14-bash-4.3-p48.drv»
```

Contents of /tmp/mytest:

```
$ cat /tmp/mytest/default.nix 
{ pkgs }:

pkgs.bash
```

It took a while to figure out that I simply had to remove the slash :-/

## Technical details

* System: 16.09.1066.417e04f (Flounder)
* Nix version: 1.11.4

### PR DESCRIPTION
**Pull request by layus**
See http://lists.science.uu.nl/pipermail/nix-dev/2016-June/thread.html#20809 for detailed issue info.

Fixes #1138.

### CODE REVIEW

**Comment by domenkozar** [comment]
cc @edolstra

File: src/libexpr/lexer.l [hunk 1]
```diff
@@ -182,8 +182,16 @@ or          { return OR_KW; }
 
 <INITIAL,INSIDE_DOLLAR_CURLY>{
 
-{PATH}      { yylval->path = strdup(yytext); return PATH; }
-{HPATH}     { yylval->path = strdup(yytext); return HPATH; }
+{PATH}      { if (yytext[yyleng-1] == '/')
+                  throw ParseError(format("Invalid path ‘%1%; trailing slashes are not allowed in paths") % yytext);
```
Comment (domenkozar) [inline src/libexpr/lexer.l]: Please don't use unicode quotes - we're trying to get rid of them.

**Comment by domenkozar** [comment]
@edolstra let's merge this? :)

**Comment by layus** [comment]
Thanks @domenkozar. I have just fixed quotes and rebased. Ready to get merged :)

File: src/libexpr/lexer.l (hunk 1 above)
Comment (layus) [inline src/libexpr/lexer.l]: Okay @ #1140 :)

**Comment by domenkozar** [comment]
@layus last thing, could you add a test?

**Comment by layus** [comment]
@domenkozar  Done !

We should somehow consider to make / a path concatenation operator, this would fix our issues here and be generally nicer.

Compare `/usr/lib + "/${var}` with `/usr/lib / var`, or even `/usr/lib/ var` if we decide to go that way :-).

**Comment by domenkozar** [comment]
@edolstra @shlevy objections?

**Pull request merged**